### PR TITLE
[SPARK-32881][CORE] Catch some race condition errors and log them more clearly 

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Decommissioning can run out of time resulting in some race condition, these race conditions result in confusing error messages but not negative impact.

### Why are the changes needed?

The NPE & element missing errors in the log can create a missunderstanding.

### Does this PR introduce _any_ user-facing change?
Logs change.

### How was this patch tested?
Existing tests pass.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
One K8s IT fails. Let's retrigger.
```
- Run in client mode. *** FAILED ***
```

**Comment by dongjoon-hyun** [comment]
Retest this please

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 1]
```diff
@@ -125,14 +125,19 @@ private class ShuffleStatus(numPartitions: Int) extends Logging {
...
+      mapStatusOpt match {
+        case Some(mapStatus) =>
+          logInfo(s"Updating map output for ${mapId} to ${bmAddress}")
+          mapStatus.updateLocation(bmAddress)
+          invalidateSerializedMapOutputStatusCache()
+        case None =>
+          logWarning(s"Asked to update map output ${mapId} for untracked map status.")
+      }
+    } catch {
+      case e: java.lang.NullPointerException =>
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:139]: Shall we remove `java.lang.`?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 2]
```diff
@@ -380,16 +380,22 @@ class BlockManagerMasterEndpoint(
...
+      rddBlocks.map { blockId =>
+        val currentBlockLocations = blockLocations.get(blockId)
+        val maxReplicas = currentBlockLocations.size + 1
+        val remainingLocations = currentBlockLocations.toSeq.filter(bm => bm != blockManagerId)
+        val replicateMsg = ReplicateBlock(blockId, remainingLocations, maxReplicas)
+        replicateMsg
+      }.toSeq
+    } catch {
+      // If the block manager has already exited, nothing to replicate.
+      case e: java.util.NoSuchElementException =>
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:396]: Shall we import?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:396]: Also, it would be great if we have a warning log here.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @holdenk ! It looks helpful. I left a few comments.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @holdenk .
Merged to master.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 1 above)
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:139]: Quick question: can we avoid catching `NullPointerException`? It's a bit odd that we catch `NullPointerException`. We could just switch to if-else I guess.

**Comment by HyukjinKwon** [comment]
How did we test this, @holdenk? I can make a quick followup for that.

**Comment by HyukjinKwon** [comment]
gentle ping. This PR introduces the very first place that catches and suppresses `NullPointerException` in the codebase - the second place that catches `NullPointerException` (`mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala`) rethrows NPE.

**Comment by HyukjinKwon** [comment]
gentile ping. Would you mind asking to address my comments? I thought it wouldn't be difficult to address.

**Comment by holdenk** [comment]
Hey sorry I didn’t see your comments. To avoid this I don’t think we could do if else, we’d have to introduce locking I think. Since this only happens under a race condition where we don’t care about the data I think it’s ok. If you’d prefer we could also take that part out since the NPE will get eaten later. Happy to review a PR with whatever approach you want or work on a follow up if you want.

**Comment by HyukjinKwon** [comment]
I can give a try. Do you have any stacktrace or symptom for this NPE issue? Seems SPARK-32881 did not throw NPE.

@dongjoon-hyun, should SPARK-32881 be resolved after this fix, or is it open mistakenly?

**Comment by dongjoon-hyun** [comment]
Thanks, @HyukjinKwon . I resolved SPARK-32881 now.

**Comment by dongjoon-hyun** [comment]
Actually, this PR fixes both `java.lang.NullPointerException` and `java.util.NoSuchElementException`.
I reported NPE to her personally in addition to SPARK-32881.

**Comment by HyukjinKwon** [comment]
That's good. I just wonder if we can just:
- `blockManagerInfo.contains(...)` instead of catching `java.util.NoSuchElementException`
-  Check if something is null, and just ignore instead of catching `java.lang.NullPointerException`

I don't believe catching `java.lang.NullPointerException` in a block is a good practice because it will catch all NPE that can potentially happen everywhere.

Would you mind if I ask the stacktrace? At least I want to understand why catching NPE is necessary even if I fail to make a fix.

**Comment by holdenk** [comment]
So I think we get the null if the mapStatus is deleted _during_ the update. The problem is checking if it's null doesn't  guarantee it won't be deleted.

**Comment by HyukjinKwon** [comment]
@holden, I am very sorry but can you guide me how you tested and/or share a traceback?

**Comment by HyukjinKwon** [comment]
@holdenk, I am very sorry but can you guide me how you tested and/or share a traceback?

**Pull request closed without merging**